### PR TITLE
[Feature panel] Show thumbnails for photo fields

### PR DESCRIPTION
**Pull request by gyarill**
issue #318

display photo on observation card with 100% width. On photo click open it in new tab

Check /project/lM8eq3EtdaY3WjzUb6Hk#f=9bWOwMlGgqiAkTVtmgpq

Screen Shot 2021-04-07 at 11 14 36 AM

### CODE REVIEW

File: web-ng/src/app/components/feature-panel/feature-panel.component.ts [hunk 1]
```diff
@@ -79,6 +80,34 @@ export class FeaturePanelComponent implements OnInit, OnDestroy {
             .pipe(map(feature => project.layers.get(feature.layerId)!))
         )
       );
+    this.photoUrls = new Map();
+    this.observations$.forEach((observations) => {
```
Comment (gino-m) [inline web-ng/src/app/components/feature-panel/feature-panel.component.ts:84]: I've never seen Observable used this way before, interesting!

File: web-ng/src/app/components/feature-panel/feature-panel.component.ts [hunk 2]
```diff
@@ -79,6 +80,34 @@ export class FeaturePanelComponent implements OnInit, OnDestroy {
...
+      observations.forEach(observation => {
+        this.getFields(observation).forEach(
+          field => {
+            if (field.type == FieldType.PHOTO) {
```
Comment (gino-m) [inline web-ng/src/app/components/feature-panel/feature-panel.component.ts:88]: Nit: These conditions can be turned into a filter()

File: web-ng/src/app/components/feature-panel/feature-panel.component.ts [hunk 3]
```diff
@@ -79,6 +80,34 @@ export class FeaturePanelComponent implements OnInit, OnDestroy {
...
+  }
+
+  openUrlInNewTab(url: string) {
+    window.open(url, '_blank');
+  }
+
+  fillPhotoURL(fieldId: string, storageFilePath: string) {
+    this.dataStoreService.getImageDownloadURL(storageFilePath)
+      .then((url) => {
+        this.photoUrls.set(fieldId, url);
```
Comment (gino-m) [inline web-ng/src/app/components/feature-panel/feature-panel.component.ts:106]: Iiuc `photoUrls` would contain all photo URLs for all fields in all observations for features shown during the lifecycle of this component. Instead, would it be easier, cleaner, and safer to instead add an optional photoUrl field to `Response`?

**Review by gino-m — CHANGES_REQUESTED** [review]
Nice work! Sorry for slow reply, I was OOO!

File: web-ng/src/app/components/feature-panel/feature-panel.component.ts (hunk 2 above)
Comment (gino-m) [reply web-ng/src/app/components/feature-panel/feature-panel.component.ts:88]: We spoke, please disregard!

File: web-ng/src/app/components/feature-panel/feature-panel.component.ts [hunk 4]
```diff
@@ -79,6 +80,34 @@ export class FeaturePanelComponent implements OnInit, OnDestroy {
             .pipe(map(feature => project.layers.get(feature.layerId)!))
         )
       );
+    this.photoUrls = new Map();
+    this.observations$.forEach((observations) => {
+      observations.forEach(observation => {
+        this.getFields(observation).forEach(
+          field => {
+            if (field.type == FieldType.PHOTO) {
+              if (observation.responses?.get(field.id)?.value as string) {
```
Comment (gino-m) [inline web-ng/src/app/components/feature-panel/feature-panel.component.ts:89]: && vs if if

File: web-ng/src/app/components/feature-panel/feature-panel.component.ts (hunk 3 above)
Comment (gino-m) [reply web-ng/src/app/components/feature-panel/feature-panel.component.ts:106]: We spoke: Please double check that feature panel is rebuilt each time a new feature is selected

Comment (gyarill) [reply web-ng/src/app/components/feature-panel/feature-panel.component.ts:106]: so Panel constructor is not called each time but this.observations$.forEach is called each time observations$
Screen Shot 2021-05-06 at 2 59 46 PM
 change

File: web-ng/src/app/components/feature-panel/feature-panel.component.ts (hunk 1 above)
Comment (gyarill) [reply web-ng/src/app/components/feature-panel/feature-panel.component.ts:84]: from the behaviour it seems that it creates subscription (?)

File: web-ng/src/app/components/feature-panel/feature-panel.component.ts (hunk 4 above)
Comment (gyarill) [reply web-ng/src/app/components/feature-panel/feature-panel.component.ts]: done

**Review by gino-m — CHANGES_REQUESTED** [review]
Thanks, @gyarill, great work! Could you please resolve lint errors/warnings so we can merge?

```

/workspace/web-ng/src/app/components/feature-panel/feature-panel.component.ts
   84:32  error  Replace `(observations)` with `observations`                                                                                                                                                                                                                                                                                                                                                                                                                                         prettier/prettier
   86:45  error  Delete `⏎··········`                                                                                                                                                                                                                                                                                                                                                                                                                                                                 prettier/prettier
   88:11  error  Replace `··if·(field.type·==·FieldType.PHOTO·&&·observation.responses?.get(field.id)?.value·as·string)·{⏎················this.fillPhotoURL(field.id,·observation.responses?.get(field.id)?.value·as·string);` with `if·(⏎············field.type·==·FieldType.PHOTO·&&⏎············(observation.responses?.get(field.id)?.value·as·string)⏎··········)·{⏎············this.fillPhotoURL(⏎··············field.id,⏎··············observation.responses?.get(field.id)?.value·as·string`  prettier/prettier
   88:28  error  Expected '===' and instead saw '=='                                                                                                                                                                                                                                                                                                                                                                                                                                                  eqeqeq
   90:13  error  Replace `··}` with `);`                                                                                                                                                                                                                                                                                                                                                                                                                                                              prettier/prettier
   91:1   error  Delete `··`                                                                                                                                                                                                                                                                                                                                                                                                                                                                          prettier/prettier
   92:9   error  Insert `}`                                                                                                                                                                                                                                                                                                                                                                                                                                                                           prettier/prettier
  102:26  error  Insert `⏎······`                                                                                                                                                                                                                                                                                                                                                                                                                                                                     prettier/prettier
  103:13  error  Replace `(url)` with `url`                                                                                                                                                                                                                                                                                                                                                                                                                                                           prettier/prettier
  106:14  error  Replace `(error)` with `error`                                                                                                                                                                                                                                                                                                                                                                                                                                                       prettier/prettier

/workspace/web-ng/src/app/components/layer-list-item/layer-list-item.component.ts
  29:10  warning  'AuthService' is defined but never used  @typescript-eslint/no-unused-vars

/workspace/web-ng/src/app/components/observation-form/observation-form.component.spec.ts
  169:9  warning  'user$' is assigned a value but never used  @typescript-eslint/no-unused-vars

/workspace/web-ng/src/app/services/data-store/data-store.service.ts
  313:2  error  Insert `·`  prettier/prettier

✖ 13 problems (11 errors, 2 warnings)
  10 errors and 0 warnings potentially fixable with the `--fix` option.
```

**Review by gino-m — APPROVED** [review]
🥳

**Pull request merged**